proto/aserto/directory/reader/v3/reader.proto

```diff
@@ -0,0 +1,350 @@
+            }
+        };
+    };
+}
+
+message GetObjectRequest {
+    // object type name (lc-string)
+    string object_type                                          = 1     [(google.api.field_behavior) = REQUIRED];
+    // object identifier (cs-string)
+    string object_id                                            = 2     [(google.api.field_behavior) = REQUIRED];
+    // materialize the object relations objects
+    optional bool with_relations                                = 3     [(google.api.field_behavior) = OPTIONAL];
+}
+
+message GetObjectResponse {
+    // object instance
+    aserto.directory.common.v3.Object result                    = 1;
+    // incoming object relations
+    repeated aserto.directory.common.v3.Relation incoming       = 2;
+    // outgoing object relations
+    repeated aserto.directory.common.v3.Relation outgoing       = 3;
+}
+
+message GetObjectManyRequest {
+    // object identifier list
+    repeated aserto.directory.common.v3.ObjectIdentifier param  = 1     [(google.api.field_behavior) = REQUIRED];
+}
+
+message GetObjectManyResponse {
+    // array of object instances
+    repeated aserto.directory.common.v3.Object results          = 1;
+}
+
+message GetObjectsRequest {
+    // object type name (lc-string)
+    string object_type                                          = 1     [(google.api.field_behavior) = OPTIONAL];
+    // pagination request
+    aserto.directory.common.v3.PaginationRequest page           = 9     [(google.api.field_behavior) = OPTIONAL];
+}
+
+message GetObjectsResponse {
+    // array of object instances
+    repeated aserto.directory.common.v3.Object results          = 1;
+    // pagination response
+    aserto.directory.common.v3.PaginationResponse page          = 9     [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+
+message GetRelationRequest {
+    // object type
+    string object_type                                          = 1     [(google.api.field_behavior) = REQUIRED];
+    // object identifier
+    string object_id                                            = 2     [(google.api.field_behavior) = REQUIRED];
+    // relation name
+    string relation                                             = 3     [(google.api.field_behavior) = REQUIRED];
+    // subject type
+    string subject_type                                         = 4     [(google.api.field_behavior) = REQUIRED];
+    // subject identifier
+    string subject_id                                           = 5     [(google.api.field_behavior) = REQUIRED];
+    // optional subject relation name
+    string subject_relation                                     = 6     [(google.api.field_behavior) = OPTIONAL];
+    // materialize relation objects
+    optional bool with_objects                                  = 7     [(google.api.field_behavior) = OPTIONAL];
+}
+
+message GetRelationResponse {
+    // array of relation instances
+    repeated aserto.directory.common.v3.Relation results        = 1;
```

Comment:
If GetRelation returns at most one relation, why do we need this to be `repeated`? It could just be `optional aserto.directory.common.v3.Relation result`.
And similarly instead of `objects` we could have two optional fields: `subject` and `object`.

Reply:
Good catch; that is a v2 leftover, will fix